### PR TITLE
VI-265: Handle malformed responses in MAP SuS service

### PR DESCRIPTION
**Pull request by dickdavis**
## Summary

This PR updates the MAP SuS configuration class to include application/json content type header and update the service to handle parsing errors.

## Related issue(s)

https://jira.devops.va.gov/browse/VI-265

## Testing done

* Start up a Rails console and require the MAP SuS service
```ruby
require 'map/sign_up/service'
```
* Instantiate a client
```ruby
client = MAP::SignUp::Service.new
```

### Status

* Make a call to the status endpoint
```ruby
client.status(icn: '1234')
```
* Verify the response is successful
```
2024-08-07 12:54:58.337053 I [54381:12260 service.rb:13] Rails -- [MAP][SignUp][Service] status success, icn: 1234
=> {:agreement_signed=>true, :opt_out=>false, :cerner_provisioned=>false, :bypass_eligible=>false}
```
* Modify the mock data response for the status call (`vets-api-mockdata/map/sign_up_service/status/default.yml`) to return invalid JSON in the response body.
* Make the call to the status endpoint again.
* Verify the response logs and raises the error
```
2024-08-07 12:55:51.203771 E [54381:12260 service.rb:17] Rails -- [MAP][SignUp][Service] status response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in `rescue in request': [MAP][SignUp][Service] status response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)
```

### Agreements accept

* Make a call to the agreements accept endpoint.
```ruby
client.agreements_accept(icn: '1234', signature_name: 'Foobar', version: '1')
```
* Verify the response is successful
```
2024-08-07 13:01:54.864865 I [54381:12260 service.rb:28] Rails -- [MAP][SignUp][Service] agreements accept success, icn: 1234
```
* Modify the mock data response for the call (`vets-api-mockdata/map/sign_up_service/agreements_accept/default.yml`) to return invalid JSON in the response body.
* Make the call again.
* Verify the response logs and raises the error
```ruby
2024-08-07 13:04:39.618259 E [54381:12260 service.rb:31] Rails -- [MAP][SignUp][Service] agreements accept response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in `rescue in request': [MAP][SignUp][Service] agreements accept response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)
```

### Agreements decline

* Make a call to the agreements decline endpoint
```ruby
client.agreements_decline(icn: '1234')
```
* Verify the response is successful
```
2024-08-07 13:09:33.370874 I [54381:12260 service.rb:39] Rails -- [MAP][SignUp][Service] agreements decline success, icn: 1234
```
* Modify the mock data response for the call (`vets-api-mockdata/map/sign_up_service/agreements_decline/default.yml`) to return invalid JSON in the response body.
* Make the call again.
* Verify the response logs and raises the error
```
2024-08-07 13:12:23.793305 E [54381:12260 service.rb:42] Rails -- [MAP][SignUp][Service] agreements decline response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in `rescue in request': [MAP][SignUp][Service] agreements decline response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)
```

### Update provisioning

* Make a call to the update provisioning endpoint
```ruby
client.update_provisioning(icn: '1234', first_name: 'foo', last_name: 'bar', mpi_gcids: 'baz')
```
* Verify the response is successful
```
2024-08-07 13:16:06.755477 I [54381:12260 service.rb:95] Rails -- [MAP][SignUp][Service] update provisioning success, icn: 1234, parsed_response: {:agreement_signed=>true, :opt_out=>false, :cerner_provisioned=>false, :bypass_eligible=>true}
=> {:agreement_signed=>true, :opt_out=>false, :cerner_provisioned=>false, :bypass_eligible=>true}
```
* Modify the mock data response for the call (`vets-api-mockdata/map/sign_up_service/update_provisioning/default.yml`) to return invalid JSON in the response body.
* Make the call again
* Verify the response logs and raises the error
```
2024-08-07 13:17:47.771446 E [54381:12260 service.rb:56] Rails -- [MAP][SignUp][Service] update provisioning response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in `rescue in request': [MAP][SignUp][Service] update provisioning response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)
```

## Screenshots

Not relevant.

## What areas of the site does it impact?

Calls to Sign-up Service.

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

None.

### CODE REVIEW

File: lib/map/sign_up/configuration.rb [hunk 1]
```diff
@@ -63,9 +63,10 @@ def connection
           request: request_options
         ) do |conn|
           conn.use :breakers
-          conn.response :betamocks if Settings.map_services.sign_up_service.mock
           conn.use Faraday::Response::RaiseError
           conn.adapter Faraday.default_adapter
+          conn.response :json, content_type: /\bjson/
```
[the remark this answers is not in the document]
Comment (dickdavis) [reply lib/map/sign_up/configuration.rb:68]: This is the standard practice for configuring a faraday client.

**Review by emilykim13 — DISMISSED** [review]
**Testing status**
After instantiating a `client` I received `2024-08-07 14:06:20.532478 I [92670:12260 service.rb:13] Rails -- [MAP][SignUp][Service] status success, icn: 1234
=> {:agreement_signed=>true, :opt_out=>false, :cerner_provisioned=>false, :bypass_eligible=>false}` as a successful result.

After modifying my mockdata, I made the call again and received `2024-08-07 14:12:43.238465 E [92670:12260 service.rb:17] Rails -- [MAP][SignUp][Service] status response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in` `rescue in request': [MAP][SignUp][Service] status response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)`. 

**Testing Agreements accept**
I made a call to the agreements accept endpoint by running `client.agreements_accept(icn: '1234', signature_name: 'Foobar', version: '1')
`. Which resulted in `2024-08-07 14:16:36.359504 I [92670:12260 service.rb:28] Rails -- [MAP][SignUp][Service] agreements accept success, icn: 1234
=> true`. 

After altering my mockdata with invalid json, I made the call again and received `2024-08-07 14:17:54.911191 E [92670:12260 service.rb:31] Rails -- [MAP][SignUp][Service] agreements accept response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in` `rescue in request': [MAP][SignUp][Service] agreements accept response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)`. 

**Testing Agreements decline**
I made the call to the agreements decline endpoint and received a successful response `2024-08-07 14:19:20.751363 I [92670:12260 service.rb:39] Rails -- [MAP][SignUp][Service] agreements decline success, icn: 1234
=> true`.

After altering my mockdata with invalid json, I made the call again and received `2024-08-07 14:20:48.090649 E [92670:12260 service.rb:42] Rails -- [MAP][SignUp][Service] agreements decline response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in` `rescue in request': [MAP][SignUp][Service] agreements decline response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)`.

**Testing Update provisioning**
I made the call to the update provisioning endpoint and received a successful response `2024-08-07 14:21:37.080823 I [92670:12260 service.rb:95] Rails -- [MAP][SignUp][Service] update provisioning success, icn: 1234, parsed_response: {:agreement_signed=>true, :opt_out=>false, :cerner_provisioned=>false, :bypass_eligible=>true}
=> {:agreement_signed=>true, :opt_out=>false, :cerner_provisioned=>false, :bypass_eligible=>true}`.

After changing my mockdata with invalid json, I received `2024-08-07 14:22:51.506324 E [92670:12260 service.rb:56] Rails -- [MAP][SignUp][Service] update provisioning response parsing error -- { :icn => "1234" }
lib/common/client/base.rb:126:in` `rescue in request': [MAP][SignUp][Service] update provisioning response parsing error, response unknown, icn: 1234 (Common::Client::Errors::ParsingError)`.


After each modification, I reverted the mockdata changes to retest for a success and received successful calls. Lgtm!

File: lib/map/sign_up/service.rb [hunk 2]
```diff
@@ -74,40 +90,43 @@ def update_provisioning_params(first_name, last_name, mpi_gcids)
       end
 
       def successful_update_provisioning_response(response, icn)
-        parsed_response = parse_response(response.body, icn, 'update provisioning')
+        parsed_response = build_response(response.body)
 
-        Rails.logger.info("#{config.logging_prefix} update provisioning success," \
+        Rails.logger.info("#{logging_prefix} update provisioning success," \
```
Comment (bosawt) [inline lib/map/sign_up/service.rb:95]:
```suggestion
        logger.info("#{logging_prefix} update provisioning success," \
```

Comment (dickdavis) [reply lib/map/sign_up/service.rb]: Good catch

**Review by bosawt — APPROVED** [review]
Emily did sufficient testing here, code looks good. Just giving the approval so you can merge now that the platform backend stuff has been resolved

**Comment by bosawt** [comment]
oh. guess it hasn't been resolved

**Comment by dickdavis** [comment]
> oh. guess it hasn't been resolved

Yeah, looks like a bug in the BE PR review check. I'm working with platform to get it figured out. I suspect it is somehow related to the check for when the code owner belongs to the team that has an exception (us, lighthouse, and mobile). Hoping to have it figured out tomorrow, but you may see some commits on this branch from us trying different things since this seems to be the first instance of that exception being used.

**Review by emilykim13 — APPROVED** [review]

**Pull request merged**